### ISSUE TITLE
bug: Cost Calculation for LiteLLM Doesn't Work

### ISSUE THREAD

**Issue by thiner** [langfuse/langfuse#3794]
### Describe the bug

I am running self-hosted Qwen2.5 model, and use LiteLLM as the proxy to serve applications. LangFuse was configured as the monitoring tool in LiteLLM. Everything is working perfectly, except the cost calculation. 
The model config:
![image](https://github.com/user-attachments/assets/264bcf80-47ff-48c9-a8b7-63f02824bec1)

The cost calculation is 0 on LangFuse.
![screenshots2024-10-17112936](https://github.com/user-attachments/assets/556ee2f8-e152-4838-a5ae-e7512a3ca7e2)

Is this a bug?

### To reproduce

LiteLLM version: `1.48.19-stable`
LLM: `Qwen2.5-72b-instruct`

### Are you interested to contribute a fix for this bug?

No

**Comment by marcklingen** [comment]
can you share a link to an example trace? happy to look into this

**Comment by thiner** [comment]
Sorry, I can't share the link. But if you need additional information, I can provide. E.g. LiteLLM configuration, other LangFuse configuration.

**Comment by thiner** [comment]
Please check below LiteLLM model configuration:

```yaml
    litellm_settings:
      drop_params: true
      success_callback: ['langfuse']
      failure_callback: ['langfuse']
      routing_strategy: 'least-busy'
    model_list:
    - litellm_params:
        api_key: "os.environ/API-KEY"
        model: openai/qwen25-72b-inst
        api_base: http://qwen25-72b-inst:8000/v1
      model_name: qwen25-72b-instruct
    - litellm_params:
        api_key: "os.environ/API-KEY"
        model: openai/qwen25-72b-inst
        api_base: http://qwen25-72b-inst:8000/v1
      model_name: qwen2.5-72b-instruct
```

**Comment by marcklingen** [comment]
Makes sense, did you add a custom model configuration in langfuse to map token usage to costs?

**Comment by thiner** [comment]
@marcklingen the model configuration as below:
![377300739-264bcf80-47ff-48c9-a8b7-63f02824bec1](https://github.com/user-attachments/assets/b6740f97-c0b9-484f-b924-003cf1dff602)

**Comment by thiner** [comment]
The previous regex pattern is incorrect for the model, I updated the regex pattern, but it's still not working. New model configuration:
![image](https://github.com/user-attachments/assets/7f8e22e5-ef58-4e63-b456-622437e5737e)

**Comment by hassiebp** [comment]
Thanks for providing the additional info @thiner ! Your regex should indeed match `qwen2.5-72b-instruct`. However, on the screenshot above, it seems like the model name is not captured correctly in the first place. Could you please share a screenshot of the generation page? You can also share a trace ID or generation ID with us to take a more detailed look.

**Comment by thiner** [comment]
@hassiebp Thie is the Metadata of generation log. I am using self-hosted Langfuse, trace ID won't help I think.
![屏幕截图 2024-10-22 101708](https://github.com/user-attachments/assets/38bce1cb-f3fd-4110-82e1-b88d71395354)

**Comment by marcklingen** [comment]
this is the trace log, does it contain a child for the generation with the respective model? a public link to the trace would help a lot

**Comment by thiner** [comment]
Is this what you want to check?
![image](https://github.com/user-attachments/assets/6014a62b-ff85-458c-842e-44ad009df757)

**Comment by hassiebp** [comment]
@thiner This looks correct to me 🤔 Are you on the latest Langfuse version? Do you see the correct cost calculation for other (e.g. OpenAI) generations?

If you are able to run SQL queries against your self-hosted Langfuse DB, the output of these two queries would help debugging this: (you can find the generation ID in the URL path)

```sql
SELECT
	calculated_total_cost
FROM
	observations
WHERE
	id = $generation_id_of_qwen_call;

SELECT
	calculated_total_cost
FROM
	observations_view
WHERE
	id = $generation_id_of_qwen_call;
```

**Comment by thiner** [comment]
Yes, when I expand the date range to 3 months, there is numbers. But the number seems incorrect, the cost of 68 million tokens should be $137 at least.
![image](https://github.com/user-attachments/assets/c1951702-9b4c-4df2-a3c5-7f170fb16bd4)

By digging into trace logs, I found the cost number actually came from one of the request, and the generation log as below:
![image](https://github.com/user-attachments/assets/39d47e9e-25d5-4889-862b-d5f4e09f2957)

The only difference here is the `stream_options: {'include_usage': True}`. Is this the root cause?

**Comment by thiner** [comment]
> @thiner This looks correct to me 🤔 Are you on the latest Langfuse version? Do you see the correct cost calculation for other (e.g. OpenAI) generations?
> 
> If you are able to run SQL queries against your self-hosted Langfuse DB, the output of these two queries would help debugging this: (you can find the generation ID in the URL path)
> 
> ```sql
> SELECT
> 	calculated_total_cost
> FROM
> 	observations
> WHERE
> 	id = $generation_id_of_qwen_call;
> 
> SELECT
> 	calculated_total_cost
> FROM
> 	observations_view
> WHERE
> 	id = $generation_id_of_qwen_call;
> ```

I ran the SQL and found that all the cost numbers are 0, except the one on screenshot above.

**Comment by hassiebp** [comment]
@thiner Thanks, this is then most likely an implementation issue in your setup. Are you on the latest Langfuse version? 

For me all looks good when reproducing locally: 

![CleanShot 2024-10-28 at 12 55 39](https://github.com/user-attachments/assets/e947a0a9-2e58-4670-9483-7100c9972097)
![CleanShot 2024-10-28 at 12 56 17](https://github.com/user-attachments/assets/a5a0951d-924d-4f2b-b1d3-5ffa24a5701a)

**Comment by thiner** [comment]
Yes, I am using the latest version `v2.86.0`.  I read this page https://langfuse.com/docs/model-usage-and-cost#compatibility-with-openai, and have setup all the fields, but the cost number is still 0. Could you please tell whether there are other fields or configuration options affecting the calculation?

**Comment by hassiebp** [comment]
No, there are not. Could you please share the result of the above pasted SQL queries when selecting all columns? I'm especially interested in the `model`, `internal_model` and `internal_model_id` columns in the `observations` table.

**Comment by thiner** [comment]
@hassiebp Here you go:
|"model"	|"internal_model"|"total_tokens"|"prompt_tokens"|"completion_tokens"|"unit"|"calculated_input_cost"|"calculated_output_cost"|"calculated_total_cost"|"internal_model_id"|
|---|---|---|---|---|---|---|---|---|---|
|"qwen25-72b-inst"|"qwen2.5"|110819|7383|	103436|"TOKENS"|	|	|0.000000000000000000000000000000|"cm2tt45md000h5ft13iup960e"|

**Comment by hassiebp** [comment]
Thanks! Now, it looks like you might have duplicate models in your models table for which the total price is set to zero somehow. The model that is selected for the cost calculation has the id `cm2tt45md000h5ft13iup960e`, could you check its prices on your models table?

```
SELECT * FROM "models" WHERE id = 'cm2tt45md000h5ft13iup960e'
```

**Comment by thiner** [comment]
It seems the model configuration is correct.
<img width="1377" alt="image" src="https://github.com/user-attachments/assets/2393815f-d841-463d-865c-1bf3bc76a4c8">

**Comment by hassiebp** [comment]
@thiner I still cannot reproduce your issue. Please make sure you are on the latest SDK + Langfuse server versions and additionally, please double check you are not sending any `update` events where you are manually setting the costs to 0 in any way. 
![CleanShot 2024-11-04 at 14 32 51](https://github.com/user-attachments/assets/4b037140-185b-4bff-9ff1-2b4949c265ab)

**Comment by luca1024-web** [comment]
> @thiner This looks correct to me 🤔 Are you on the latest Langfuse version? Do you see the correct cost calculation for other (e.g. OpenAI) generations?
> 
> If you are able to run SQL queries against your self-hosted Langfuse DB, the output of these two queries would help debugging this: (you can find the generation ID in the URL path)
> 
> ```sql
> SELECT
> 	calculated_total_cost
> FROM
> 	observations
> WHERE
> 	id = $generation_id_of_qwen_call;
> 
> SELECT
> 	calculated_total_cost
> FROM
> 	observations_view
> WHERE
> 	id = $generation_id_of_qwen_call;
> ```

I encountered the same problem. According to your suggestion, I made the corresponding query and found that calculated_total_cost is NULL. And I created related issues in [dify.](https://github.com/langgenius/dify/issues/10253) Can you spare your precious time to help me locate where the problem is?
![image](https://github.com/user-attachments/assets/7f038712-51c3-44d7-8646-a6131b2f5896)

**Comment by raihan0824** [comment]
got the same problem, do you know why it occurs?

**Comment by MarcinRybinski** [comment]
I've encountered the same issue and the problem was on LiteLLM site.
LiteLLM is calculating costs on his side and when providing configuration to your own custom model you also have to pass input and output costs:

```
- model_name: your_model
  litellm_params:
    ...
    input_cost_per_token: 0.123456
    output_cost_per_token: 0.234567
```

If you won't provide the cost:
1. LiteLLM assumes that the input and output costs are 0
2. When they're 0, the total cost is of cource = 0
3. When you take a look at [LangFuse integration code in LiteLLM repo](https://github.com/BerriAI/litellm/blob/main/litellm/integrations/langfuse/langfuse.py):
    1. [The cost is taken from response_cost](https://github.com/BerriAI/litellm/blob/main/litellm/integrations/langfuse/langfuse.py#L579C13-L579C53) field which will be 0 (because you didn't provide input and output cost)
    2. Then the cost (which is 0) [is saved in usage object as total_cost](https://github.com/BerriAI/litellm/blob/main/litellm/integrations/langfuse/langfuse.py#L652-L656)
    3. [Then this usage object is used to create params which later on are passed to langfuse.trace](https://github.com/BerriAI/litellm/blob/main/litellm/integrations/langfuse/langfuse.py#L679-L692)
4. In that way, we are sending the request to the LangFuse with total_cost = 0
5. When the Langfuse **worker** receives request, [it has a logic that if the requests contains the cost information, just use this cost](https://github.com/langfuse/langfuse/blob/main/worker/src/services/IngestionService/index.ts#L719-L745) - so in this case, LangFuse doesn't calculates the costs at all.
6. That's why we see the cost = 0, because request to the langfuse contains the total_cost = 0.

As far as I studied LiteLLM documentation, there's no possibility to disable calculating the costs by LiteLLM, so we have to provide cost information in two places - in LiteLLM which calculates the cost and in LangFuse UI for documentation purposes.

**Comment by marcklingen** [comment]
Agree that for models without cost litellm should not set cost to 0 in Laangfuse, do you think this could be a quick patch of the litellm integration?
